_posts/2015-04-17-Enums-match-mutation-and-moves.md

```diff
@@ -0,0 +1,740 @@
+---
+layout: post
+title: "Enums: match, mutation, and moves"
+author: Felix S. Klock II
+description: "A tour of enums in Rust."
+---
+
+In this post we explore one corner of the Rust language: defining a
+class of data via `enum`, and processing instances of such data via
+`match`.
+
+Most of the post will use a single running example: modeling a game of
+chess. At the end the post changes gears, shifting to a discussion of
+how affine-typing ensures that the demonstrated features do not
+introduce unsoundness.
+
+## Tutorial material
+
+The post takes the overall form of a tutorial.
+
+If you are already familiar with defining `enums` and simple uses of
+`match`, you may want to skip ahead to the [mutating enums] or
+[avoiding unsoundness] sections. This first section is meant to
+establish everything one would need to know about enums to understand
+the discussion in latter two sections.
+
+[mutating enums]: #mutating-enums
+[avoiding unsoundness]: #avoiding-unsoundness
+
+### Defining simple enums
+
+The pieces of chess can be classified as follows:
+
+```rust
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum PieceShape {
+    King, Queen, Rook, Bishop, Knight, Pawn,
+}
+```
+
+The `derive` line automatically generates support for three common
+operations: the `Copy, Clone` gives us the ability to freely copy or
+clone instances of `PieceShape`, the `Debug` provides support for
+printing each piece-rank into an output buffer, e.g. via
+`println!("piece: {:?}", piece)`, and `PartialEq, Eq` provides support
+for the `==` operator on instances of the enum.
```

Comment:
Perhaps the brief explanations here should be worded a bit differently, something like this?

> The `derive` line automatically generates support for three common operations: `Copy` and `Clone` give us the ability to freely copy or clone an instance of `PieceShape`; `Debug` allows us to print each piece-rank into an output buffer, e.g. via `println!("piece: {:?}", piece)`; and finally `PartialEq` and `Eq` enable use of the `==` operator on instances of `PieceShape`.

Also, I can see that you are trying to answer 'what the heck is this derive thing??' without having to explain attributes (or even `derive` specifically) and traits, but maybe they should at least be identified as such with some sort of note about the full explanation being beyond the scope of this post:

> `derive` is an attribute that tells the compiler to insert generated implementations of the traits we specified. Traits are **bla bla bla** and do **some things**.  We won't cover the details of attributes and traits in this tutorial, but for more information you can read the following sections of *The Rust Programming Language*: [Attributes](http://doc.rust-lang.org/nightly/book/attributes.html), [Traits](http://doc.rust-lang.org/nightly/book/traits.html).